### ISSUE TITLE
How to copy a blob efficiently? Is start_copy(const cloud_blob &source) sync?

### ISSUE THREAD

**Issue by TiTi** [Azure/azure-storage-cpp#137]
Hi,

**TL;DR:**
Does `utility::string_t start_copy(const cloud_blob& source)` wait for end of copy?

-----

The documentation is not clear enough how to copy efficiently a blob to another one. It says: 

> utility::string_t azure::storage::cloud_blob::start_copy	(	const cloud_blob & 	source	)
> Begins an operation to copy a blob's contents, properties, and metadata to a new blob.

Thus, as the method name suggest, the copy is starting, but it might not be yet finished?
This idea is confirmed when looking at the azure-storage-cpp tests, cf [Microsoft.WindowsAzure.Storage/tests/cloud_blob_test.cpp](https://github.com/Azure/azure-storage-cpp/blob/54145423f12e0b7cedc15824b506e2660321507e/Microsoft.WindowsAzure.Storage/tests/cloud_blob_test.cpp) : 
* just after `.start_copy()` is called, you call a helper `wait_for_copy()`
* [this helper](https://github.com/Azure/azure-storage-cpp/blob/54145423f12e0b7cedc15824b506e2660321507e/Microsoft.WindowsAzure.Storage/tests/cloud_blob_test.cpp#L27) does a loop until copy status changes from azure::storage::copy_status::pending ; At each try:
  * the blob properties are refreshed with `.download_attributes()`
  * there is a **1 second sleep**, ouch!

Therefore, in my original implementation, it did similarly to your `wait_for_copy()` code.
But the 1 second perf. is really bad for me performance wise.
So I reduced the sleep to a 100ms wait. And then... while trying to get rid of the first sleep to optimize things, I've discovered that sleep is unecessary.
Whatever blob size I'm copying: 10Ko file, 60Mo or 750Mo file, it does not need to loop.
Here is the code i used:

```
azure::storage::cloud_block_blob existingblob = _az_container.get_block_blob_reference(pathFrom);
if (existingblob.exists())
{
	azure::storage::cloud_block_blob newblob = _az_container.get_block_blob_reference(pathTo);
	if (!newblob.exists())
	{
		newblob.start_copy(existingblob);
		int count = 0;
		do
		{
			count++;

			/*
			std::this_thread::sleep_for(std::chrono::milliseconds(100));
			newblob.download_attributes();
			if (count > 600)
			{
				SetError("Copy time-out", ERROR_NET_WRITE_FAULT);
				return FALSE;
			}*/
		} while (newblob.copy_state().status() == azure::storage::copy_status::pending);

		// count == 1 here, whatever the blob size (10Ko file, 60Mo or 750Mo)

		return newblob.copy_state().status() == azure::storage::copy_status::success;
	}
	...
```

I initially tried with **Azure Storage Emulator**, but I get the same result with **a real azure storage account**.

The sleep doesn't seem necessary because count always equals 1 meaning the copy_status is never **pending**, at least after my call to start_copy().

The copy is quite fast even with a 750Mo file [it doesn't look like the file is fully copied underneath, which is super awesome performance-wise].

When going to definition (F12) of the start_copy() i'm using, i end up to this:
```
        utility::string_t start_copy(const cloud_blob& source)
        {
            return start_copy_async(source).get();
        }
```
Which is confusing because it's using async code **and wait** for the end of operation.
Yet the name is still **start_copy**_async and as the code documentation says:
> Begins an operation to copy a blob's contents, properties, and metadata to a new blob.

Thus it's not waiting for end of copy?

It's pretty clear in the [azure::storage::cloud_blob Class Reference](http://azure.github.io/azure-storage-cpp/classazure_1_1storage_1_1cloud__blob.html) that most methods have an `_async` version.

For instance `download_to_file` and `upload_from_file` wait for end of download/upload, but the `_async` version do not, as expected.

Therefore, my questions are simple : 
* Does `utility::string_t start_copy(const cloud_blob& source)` wait for end of copy?
* Why the name is `start_copy()` and not `copy()` with also `copy_async()`? 
* Why do you have a 1 second sleep in your tests ?
* Can I remove the loop and sleep from my code to wait for end of copy?

--------------

By the way typo in blob.h : Intitiates -> initiates (150 matches!)

**Comment by JasonYang-MSFT** [comment]
Hi @TiTi ,

To answer your questions:
1. utility::string_t start_copy(const cloud_blob& source) won't wait for the end of copy.
2. start copy means once this api called, we will kick off a copy on server side. We can't grantee this copy is finished after this api call, because copy is possibly still ongoing on server side.
3. Copy may take some time on server side depends on the server load. Also, cross region copy may take more time because of the network limit.
4. Yes. If start_copy() returned with no exceptions, copy is scheduled on server side, loop or sleep won't affect the copy. A bit advise, you can first check the copy status after start_copy(). Usually the copy will finish fast and directly return success in start_copy() if the source and destination are in the same region. 
Also, if there are lots of blob need to be copied, you kick off copies in one time and don't need to wait for the previous one to finish.

**Comment by TiTi** [comment]
Ok thanks.
